File: src/ios/SignicatPlugin.swift
import Foundation
import ConnectisSDK
import UIKit



@objc(SignicatPlugin)
class SignicatPlugin: CDVPlugin, AuthenticationResponseDelegate, AccessTokenDelegate  {

    private var currentCommand: CDVInvokedUrlCommand?
    private var accessTokenCallbackId: String?

    
    override func pluginInitialize() {
        NotificationCenter.default.addObserver(
          self,
          selector: #selector(continueUserActivityHandler(_:)),
          name: Notification.Name(rawValue: "UIApplicationContinueUserActivity"),
          object: nil
        )
    }

    @objc(continueUserActivityHandler:) func continueUserActivityHandler(_ notification: NSNotification) {
        let userActivity = notification.object as! NSUserActivity
        if ConnectisSDK.continueLogin(userActivity: userActivity) {
            NSLog("[Signicat] continueLogin handled the URL")
        }
    }


    /**
    * Requests an OpenID Connect access token from the Signicat Identity Broker.
    *
    * On iOS the Mobile SDK exposes an API that returns a valid OpenID access
    * token once the user has previously authenticated. This access token
    * represents an OAuth2 authorization credential that you can use to call
    * backend services on behalf of the authenticated user. Treat this token
    * as a secret and never expose it to untrusted components.
    *
    */

    @objc(getAccessToken:)
    @MainActor
    func getAccessToken(command: CDVInvokedUrlCommand) {

        self.accessTokenCallbackId = command.callbackId

        ConnectisSDK.useAccessToken(
            caller: self.viewController,
            delegate: self
        )

    }

    func handleAccessToken(accessToken: Token) {

        let result = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: accessToken.getValue()
        )
        self.commandDelegate.send(
            result,
            callbackId: self.accessTokenCallbackId
        )

    }

    func onError(errorMessage: String) {
        guard let callbackId = self.accessTokenCallbackId else { return }
        sendError(code:"E_ACCESS_TOKEN_EXCEPTION", message:errorMessage, callbackId: callbackId)
    }



    /**
    * Initiates the Signicat Identity Broker authentication flow.
    *
    * On iOS the Mobile SDK’s `login` method takes a configuration object
    * containing issuer, client ID, redirect URI and optional scopes.
    * It opens a browser or app-to-app flow depending on configuration,
    * and calls back into delegate methods when complete.
    *
    * The SDK supports:
    *   • A WEB login flow — browser-based authentication with universal links.
    *   • An APP_TO_APP flow — universal linking to an external identity app (Digid).
    *
    * The results are sent asynchronously:
    *   • On success, `handleResponse(...)` receives an AuthenticationResponse.
    *   • On cancel, `onCancel()` is invoked.
    *   • SDK errors are returned via an error handler.
    *
    */

    @objc(loginAppToApp:)
    @MainActor
    func loginAppToApp(command: CDVInvokedUrlCommand) {

        NSLog("loginAppToApp Started");

        self.currentCommand = command

        guard command.arguments.count >= 5,
            let issuer = command.arguments[0] as? String,
            let clientID = command.arguments[1] as? String,
            let redirectURI = command.arguments[2] as? String,
            let appToAppScopes = command.arguments[3] as? String,
            let brokerDigidAppAcs = command.arguments[4] as? String,
            let isAppToApp = command.arguments[5] as? Bool
        else {
            sendError(code:"E_LOGIN_INVALID_ARGS", message:"Missing or invalid parameters", callbackId: command.callbackId)
            return
        }

        let configuration: ConnectisSDKConfiguration
        do {
            configuration = ConnectisSDKConfiguration(
                issuer: issuer,
                clientID: clientID,
                redirectURI: redirectURI,
                scopes: appToAppScopes,
                brokerDigidAppAcs: brokerDigidAppAcs,
                loginFlow: isAppToApp ? LoginFlow.APP_TO_APP : LoginFlow.WEB
            )
        } catch {
            sendError(code:"E_LOGIN_CONFIG", message:"Invalid login configuration: \(error.localizedDescription)", callbackId: command.callbackId)
            return
        }


        ConnectisSDK.logIn(
            sdkConfiguration: configuration,
            caller: self.viewController,
            delegate: self,
            allowDeviceAuthentication: false//ConnectisSDK.isDeviceAuthenticationEnabled()
        )
    }


    func handleResponse(authenticationResponse: AuthenticationResponse) {
        NSLog("loginAppToApp handleResponse");
        guard let command = currentCommand else { return }

        do{

            var json: [String: Any] = ["isSuccess": authenticationResponse.isSuccess]
            json["error"] = authenticationResponse.error ?? NSNull()
            json["nameIdentifier"] = authenticationResponse.nameIdentifier ?? NSNull()
            var attributesJson: [[String: Any]] = []
            for attr in authenticationResponse.attributes ?? [] {
                attributesJson.append([
                    "name": attr.name,
                    "value": attr.value
                ])
            }
            json["attributes"] = attributesJson

            // Convert to JSON string safely
            var jsonString = "{}"
            if JSONSerialization.isValidJSONObject(json),
            let data = try? JSONSerialization.data(withJSONObject: json, options: []) {
                jsonString = String(data: data, encoding: .utf8) ?? "{}"
            }

            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: jsonString
            )

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            self.currentCommand = nil
            
        } catch {
            sendError(code:"E_HANDLE_RESPONSE_EXCEPTION", message:"Error handling login response: \(error)", callbackId: command.callbackId)
            return
        }
        
    }



    func onCancel() {
        guard let command = currentCommand else { return }

        sendError(code:"E_LOGIN_CANCELED", message:"User canceled login", callbackId: command.callbackId)

        self.currentCommand = nil
    }


    /**
    * Sends a structured error response back to the Cordova layer.
    *
    * This helper method wraps an error code and message into a JSON object,
    * serializes it into a JSON string, and returns it through a Cordova
    * `CDVPluginResult` with `CDVCommandStatus_ERROR`.
    *
    * The JSON structure returned to JavaScript has the form:
    * {
    *   "code": "<error-code>",
    *   "message": "<human-readable description>"
    * }
    *
    * This method is used to standardize native iOS error reporting for:
    *   • Login flow failures
    *   • Access token retrieval errors
    *   • Invalid arguments or configuration issues
    *   • SDK-level exceptions from Signicat Mobile SDK
    *
    * On the JavaScript side, this allows consistent handling of all errors
    * with predictable fields for logging or user-facing messages.
    *
    * - Parameters:
    *   - code: A short machine-readable error identifier (e.g. `"LOGIN_FAILED"`).
    *   - message: A descriptive human-readable explanation of the error.
    *   - callbackId: The Cordova callback ID associated with the original request.
    */

    func sendError(code: String, message: String, callbackId: String) {
        let errorObj: [String: Any] = [
            "code": code,
            "message": message
        ]

        // Convert to JSON string safely
        var jsonString = "{}"
        if JSONSerialization.isValidJSONObject(errorObj),
        let data = try? JSONSerialization.data(withJSONObject: errorObj, options: []) {
            jsonString = String(data: data, encoding: .utf8) ?? "{}"
        }

        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: jsonString
        )

        self.commandDelegate.send(
            pluginResult, 
            callbackId: callbackId
        )
    }

}

File: src/android/SignicatPlugin.java
package com.signicat.plugin;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import com.google.gson.Gson;

import com.connectis.sdk.ConnectisSDK;
import com.connectis.sdk.api.configuration.ConnectisSDKConfiguration;
import com.connectis.sdk.api.authentication.AuthenticationResponse;
import com.connectis.sdk.api.authentication.AuthenticationResponseDelegate;
import com.connectis.sdk.api.authentication.ErrorResponseDelegate;
import com.connectis.sdk.internal.authentication.login.LoginFlow;
import com.connectis.sdk.api.authentication.AccessTokenDelegate;
import com.connectis.sdk.api.authentication.Token;
import org.jetbrains.annotations.NotNull;


/*********************************************************************************
 * The Android SDK is built with AndroidX. 
 * This means that you must have an AndroidX application to use the Android SDK.
 * The supported Android minimum version is Android 9 (API level 28). 
 * The supported Android target version is Android 16 (API level 36).
 * The SDK is made using Kotlin 1.9. If your app uses Kotlin, then we recommend that you use a language version that is the same (1.9) or newer.
 * The target phone's default browser must support cookies. If not, then a browser that supports cookies must be set as default. 
 * 
 * It includes the android-sdk-1.1.7.aar library.
 * 
**********************************************************************************/

public class SignicatPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if ("loginAppToApp".equals(action)) {
            login(args, callbackContext);
            return true;
        }

        if ("getAccessToken".equals(action)) {
            getAccessToken(callbackContext);
            return true;
        }

        return false;
    }


    /**
     * Requests a valid OpenID access token from the Signicat Identity Broker.
     *
     * This method initiates an SDK call that returns an OAuth2/OpenID access token
     * previously obtained through a successful login. The token can be used to
     * authorize calls to backend APIs that require a valid access credential.
     *
     * The AccessTokenDelegate provides two callbacks:
     *   - handleAccessToken(Token accessToken): called when a valid token is available.
     *   - onError(String exception): called when there was a problem obtaining
     *     the token (e.g., expired session, internal error).
     *
     * Note: Treat the returned access token as a secret and keep it secure.
     */

    private void getAccessToken(final CallbackContext callbackContext) {    

        ConnectisSDK.Companion.useAccessToken(
            cordova.getActivity(),
            new AccessTokenDelegate() {
                @Override
                public void handleAccessToken(@NotNull Token accessToken) {
                    callbackContext.success(accessToken.getValue());
                }

                @Override
                public void onError(@NotNull String exception) {
                    sendError(callbackContext, "E_ACCESS_TOKEN_EXCEPTION", exception);
                }
            }
        );

    }

    /**
     * Starts the authentication flow with the Signicat Identity Broker.
     *
     * This method uses ConnectisSDKConfiguration to launch either:
     *   - a WEB login flow using an external browser,
     *   - or an APP_TO_APP flow for DigID/App-to-app redirect authentication.
     *
     * After the user interacts with the broker (e.g., entering credentials,
     * choosing an ID provider), the result is returned through:
     *   - AuthenticationResponseDelegate.handleResponse(AuthenticationResponse):
     *     called with a response object when authentication completes.
     *   - AuthenticationResponseDelegate.onCancel():
     *     called if the user cancels the login.
     *
     * If an error occurs during the login process (e.g., network error,
     * invalid configuration), ErrorResponseDelegate.handleError(Exception e)
     * is triggered with details of the failure.
     *
     * The response includes information such as:
     *   - success status,
     *   - unique name identifier,
     *   - attributes returned by the identity provider (e.g., subject, amr).
     */

    private void login(JSONArray args, CallbackContext callbackContext) {

        final String issuer;
        final String clientId;
        final String redirectUri;
        final String scopes;
        final String brokerDigidAppAcs;
        final boolean isAppToApp;

        try {
            issuer = args.getString(0);
            clientId = args.getString(1);
            redirectUri = args.getString(2);
            scopes = args.getString(3);
            brokerDigidAppAcs = args.getString(4);
            isAppToApp = args.optBoolean(5, false);

        } catch (JSONException e) {
            sendError(callbackContext, "E_LOGIN_INVALID_ARGS", e.getMessage());
            return;
        }

        final boolean allowDeviceAuthentication = false;

        cordova.getActivity().runOnUiThread(() -> {
          try {

            ConnectisSDKConfiguration configuration = new ConnectisSDKConfiguration(
                issuer,
                clientId,
                redirectUri,
                scopes,
                null,
                brokerDigidAppAcs,
                (isAppToApp) ? LoginFlow.APP_TO_APP : LoginFlow.WEB
            );
        
            AuthenticationResponseDelegate delegate = new AuthenticationResponseDelegate() {
              @Override
              public void handleResponse(AuthenticationResponse response) {
                Gson gson = new Gson();
                String responseJSON = gson.toJson(response);
                callbackContext.success(responseJSON);
              }
        
              @Override
              public void onCancel() {
                sendError(callbackContext, "E_LOGIN_CANCELED", "User canceled Signicat login");
              }
            };
        
            ErrorResponseDelegate errorDelegate = new ErrorResponseDelegate() {
              @Override
              public void handleError(Exception e) {
                sendError(callbackContext, "E_LOGIN_SDK_ERROR", e.getMessage());
              }
            };
        
            ConnectisSDK.Companion.login(
                configuration,
                cordova.getActivity(),
                delegate,
                errorDelegate,
                allowDeviceAuthentication
            );

          } catch (Exception e) {
            sendError(callbackContext, "E_LOGIN_EXCEPTION", e.getMessage());
          }
        });

    }

    /**
     * Sends a structured error object back to the Cordova JavaScript layer.
     *
     * This method wraps an error code and message into a JSON object,
     * then passes it to the Cordova CallbackContext as an error response. The
     * resulting JSON has the following structure:
     *
     * {
     *   "code": "<ERROR_CODE>",
     *   "message": "<Human readable message>"
     * }
     *
     * Using a consistent JSON error format allows the JavaScript side to reliably
     * detect, classify, and handle plugin errors such as argument validation issues,
     * SDK failures, login cancellations, or unexpected exceptions.
     *
     * @param ctx     The Cordova CallbackContext used to report the error.
     * @param code    A short, machine-readable error code (e.g. "E_LOGIN_CONFIG").
     * @param message A descriptive message providing additional context.
     */
    
    private void sendError(CallbackContext ctx, String code, String message) {
        try {
            JSONObject errorJson = new JSONObject();
            errorJson.put("code", code);
            errorJson.put("message", message);
            ctx.error(errorJson.toString());
        } catch (JSONException ignored) {
            ctx.error("{\"code\":\"" + code + "\",\"message\":\"" + message + "\"}");
        }
    }
}

